pkg/storegateway/bucket.go

```diff
@@ -646,36 +719,54 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		})
 
 		for seriesSet.Next() {
-			var lset labels.Labels
-			var series storepb.Series
-
-			seriesCount++
-
 			// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
 			// because the subsequent call to seriesSet.Next() may release it.
-			if req.SkipChunks {
-				lset, _ = seriesSet.At()
-			} else {
-				lset, series.Chunks = seriesSet.At()
+			lset, chks := seriesSet.At()
+			seriesCount++
+			msg := &grpc.PreparedMsg{}
+			if req.StreamingChunksBatchSize > 0 && !req.SkipChunks {
+				// We only need to stream chunks here because the series labels have already
+				// been sent above.
+				// TODO: is the 'is end of stream' parameter required here?
+				streamingChunks := storepb.StreamSeriesChunks{
+					SeriesIndex: uint64(seriesCount - 1),
+					Chunks:      chks,
+				}
 
-				chunksCount += len(series.Chunks)
-				s.metrics.chunkSizeBytes.Observe(float64(chunksSize(series.Chunks)))
+				// Encode the message. We encode it ourselves into a PreparedMsg in order to measure
+				// the time it takes.
+				encodeBegin := time.Now()
+				if err = msg.Encode(srv, storepb.NewStreamSeriesChunksResponse(&streamingChunks)); err != nil {
```

Comment:
I know this was what the existing code was doing, but should we send more than one series per message if the total size of chunks is less than \<some limit>? (I don't think this is something we have to change in this PR, might be something we can improve in the future.)

Reply:
That change will require an update in the proto. So if we want to do it, we should do it now. I can add that.

Reply:
I didn't look at the grpc buffer size. But looking at the recent test results, grpc buffer limits seems to be doing its thing. One advantage of batching together chunks of different series here is we potentially save on a lot of grpc overhead on CPU.